File: modules/my-battery/android/src/main/java/expo/modules/mybattery/MyBatteryModuleSharedObject.kt
package expo.modules.mybattery

import expo.modules.kotlin.AppContext
import expo.modules.kotlin.sharedobjects.SharedObject

class MyBatteryModuleSharedObject(appContext: AppContext) : SharedObject(appContext) {
  var count: Int = 0

  override fun sharedObjectDidRelease() {
    super.sharedObjectDidRelease()
  }
}

File: modules/my-battery/src/MyBatteryModuleSharedObject.ts
import { SharedObject, useReleasingSharedObject } from 'expo-modules-core';

import MyBatteryModule from './MyBatteryModule';

export declare class MyBatteryModuleSharedObject extends SharedObject {
  count: number;
}

/**
 * Creates a new MyBatteryModuleSharedObject instance.
 * You are responsible for releasing it from memory by calling `release()` when done.
 */
export function createMyBatteryModuleSharedObject(): MyBatteryModuleSharedObject {
  return new MyBatteryModule.MyBatteryModuleSharedObject();
}

/**
 * A hook that creates a MyBatteryModuleSharedObject instance and automatically
 * releases it when the component unmounts.
 */
export function useMyBatteryModuleSharedObject(): MyBatteryModuleSharedObject {
  return useReleasingSharedObject(() => new MyBatteryModule.MyBatteryModuleSharedObject(), []);
}
